File: ahrs_ekf.py
import numpy as np

class AhrsEKF:

    mag_cal_A = 0 # TODO
    mag_cal_b = 0

    def __init__(self, A, b):
        self.mag_cal_A = A
        self.mag_cal_b = b

    def initialize(self, accel_init, mag_init):
        # create initial rotation matrix
        C1 = np.cross(np.cross(accel_init, mag_init), accel_init)
        C1 = C1/np.linalg.norm(C1)
        C2 = np.cross(accel_init, mag_init)
        C2 = C2/np.linalg.norm(C2)
        C3 = accel_init
        C3 = C3/np.linalg.norm(C3)
        self.C_init = np.column_stack((C1, C2, C3))

        q_hat1 = (1/2)*np.sqrt(self.C_init[0][0]+self.C_init[1][1]+self.C_init[2,2]+1)
        q_hat2 = (1/2)*np.sign(self.C_init[2][1]-self.C_init[1][2])*np.sqrt(self.C_init[0][0]-self.C_init[1][1]-self.C_init[2][2]+1)
        q_hat3 = (1/2)*np.sign(self.C_init[0][2]-self.C_init[2][0])*np.sqrt(self.C_init[1][1]-self.C_init[2][2]-self.C_init[0][0]+1)
        q_hat4 = (1/2)*np.sign(self.C_init[1][0]-self.C_init[0][1])*np.sqrt(self.C_init[2][2] - self.C_init[0][0]-self.C_init[1][1]+1)
        self.q_hat = np.vstack((q_hat1, q_hat2, q_hat3, q_hat4))
        
        self.P = 10*np.identity(4)

        # set tunings
        sigma_gyro = 0.075
        self.Sigma_gyro = (sigma_gyro**2)*np.identity(3)

        sigma_acc = 0.5 # TODO: change tunings
        sigma_mag = 1.1
        self.R = np.identity(6)
        self.R[0][0] = (sigma_acc**2)
        self.R[1][1] = (sigma_acc**2)
        self.R[2][2] = (sigma_acc**2)
        self.R[3][3] = (sigma_mag**2)
        self.R[4][4] = (sigma_mag**2)
        self.R[5][5] = (sigma_mag**2)

        magnetic_dip_angle = np.deg2rad(-14.78)
        self.magnetic_field_vec = np.array([np.cos(magnetic_dip_angle), 0, np.sin(magnetic_dip_angle)]).T/np.sqrt(np.cos(magnetic_dip_angle)**2 + np.sin(magnetic_dip_angle)**2)
        self.grav_vec = np.array([0, 0, -1]).T

        print("AHRS EKF Initialized")

    def time_update(self, gyro, dt):
        omega_skew_sym = vec3ToSkewSym(gyro)
        Omega_row1 = np.hstack(([0], -gyro.T))
        Omega_block2 = np.column_stack((gyro.T, vec3ToSkewSym(gyro)))
        Omega = np.vstack((Omega_row1, Omega_block2))

        q_w_a = np.cos(np.linalg.norm(gyro)*dt/2)
        q_w_v = np.sin(np.linalg.norm(gyro)*dt/2)*gyro/np.linalg.norm(gyro)
        q_w = np.hstack((q_w_a, q_w_v)).T
        self.q_hat = quatmult(self.q_hat, q_w)
        self.q_hat = self.q_hat[0]

        self.A_d = np.array([[1, (-dt/2)*gyro[0], (-dt/2)*gyro[1], (-dt/2)*gyro[2]],
                             [(dt/2)*gyro[0], 1, (dt/2)*gyro[2], (-dt/2)*gyro[1]],
                             [(dt/2)*gyro[1], (-dt/2)*gyro[2], 1, (dt/2)*gyro[0]],
                             [(dt/2)*gyro[2], (dt/2)*gyro[1], (-dt/2)*gyro[0], 1]])
        # Jacobian of f w.r.t. angular rates (omega)
        W = (dt/2)*np.array([[-self.q_hat[1],  -self.q_hat[2],  -self.q_hat[3]],
                             [self.q_hat[0],   -self.q_hat[3],   self.q_hat[2]],
                             [self.q_hat[3],    self.q_hat[0],  -self.q_hat[1]],
                             [-self.q_hat[2],   self.q_hat[1],  -self.q_hat[0]]])
        Q = np.matmul(np.matmul(W,self.Sigma_gyro),W.T)

        P = np.matmul(np.matmul(self.A_d, self.P), self.A_d.T) + Q

    def measurement_update(self, accel, mag):
        accel_normalized = accel/np.linalg.norm(accel)
        mag_normalized = mag/np.linalg.norm(mag)

        rot_mat = quat2rot(self.q_hat)
        accel_hat = np.matmul(rot_mat.T,self.grav_vec.T)
        mag_hat = np.matmul(rot_mat.T,self.magnetic_field_vec.T)
        Y_hat = np.hstack((accel_hat, mag_hat))

        u_g = np.cross(self.grav_vec, self.q_hat[1:])
        u_r = np.cross(self.magnetic_field_vec, self.q_hat[1:])
        skew_sym_first_row = vec3ToSkewSym(u_g + self.q_hat[0]*self.grav_vec)
        skew_sym_second_row = vec3ToSkewSym(u_r + self.q_hat[0]*self.magnetic_field_vec)
        H_12_block = skew_sym_first_row + np.dot(self.q_hat[1:], self.grav_vec)*np.identity(3) - np.matmul(self.grav_vec, self.q_hat[1:].T)
        H_22_block = skew_sym_second_row +np.dot(self.q_hat[1:], self.magnetic_field_vec)*np.identity(3) - np.matmul(self.magnetic_field_vec, self.q_hat[1:].T)
        H_row1_block = np.hstack((u_g[:, None], H_12_block))
        H_row2_block = np.hstack((u_r[:, None], H_22_block))
        H = 2*np.vstack((H_row1_block, H_row2_block))

        Y = np.hstack((accel_normalized, mag_normalized))

        # compute Kalman gain
        v = Y - Y_hat
        S = np.matmul(np.matmul(H, self.P), H.T) + self.R
        K = np.matmul(np.matmul(self.P, H.T), np.linalg.inv(S))

        self.q_hat = self.q_hat + np.matmul(K, v)
        self.P = np.matmul((np.identity(4) - np.matmul(K, H)), self.P.T)

        print(self.q_hat)
        print(self.P)

def vec3ToSkewSym(q):
    skew_sym_mat = np.array([[0, -q[2], q[1]],
                             [q[2], 0, -q[0]],
                             [-q[1], q[0], 0]])
    return skew_sym_mat

def quatmult(q0,q1):
   # quaternion multiplication
    
   q0_q1_w = q0[0]*q1[0] - q0[1]*q1[1] - q0[2]*q1[2] - q0[3]*q1[3]
   q0_q1_x = q0[0]*q1[1] + q0[1]*q1[0] + q0[2]*q1[3] - q0[3]*q1[2]
   q0_q1_y = q0[0]*q1[2] - q0[1]*q1[3] + q0[2]*q1[0] + q0[3]*q1[1]
   q0_q1_z = q0[0]*q1[3] + q0[1]*q1[2] - q0[2]*q1[1] + q0[3]*q1[1]
   
   q0_q1 = np.array([q0_q1_w, q0_q1_x, q0_q1_y, q0_q1_z]).T;    
   
   return q0_q1


# converted from Howard Chen's matlab function
def quat2rot(q):
   #  quatToDCM converts from quaternion vector to DCM
   #  input: quaternion vector [nx3] first element is real
   #  output: Rotation matrix [2x2xn]
   #
   #  Author: not Howard Chen


   R = np.zeros((3,3))


   R[0, 0] = q[0]**2+q[1]**2-q[2]**2-q[3]**2
   R[0, 1] = -2*q[0]*q[3]+2*q[1]*q[2]
   R[0, 2] = 2*q[0]*q[2]+2*q[1]*q[3]
   R[1, 0] = 2*q[0]*q[3]+2*q[1]*q[2]
   R[1, 1] = q[0]**2-q[1]**2+q[2]**2-q[3]**2
   R[1, 2] = -2*q[0]*q[1]+2*q[2]*q[3]
   R[2, 0] = -2*q[0]*q[2]+2*q[1]*q[3]
   R[2, 1] = 2*q[0]*q[1]+2*q[2]*q[3]
   R[2, 2] = q[0]**2-q[1]**2-q[2]**2+q[3]**2

   return R


# A = np.identity(3)
# b = np.array([-0.2061, -0.0178, -0.0520])
# a = AhrsEKF(A, b)
# mag_init = np.array([0.0542322857859514, -0.215706835372906, -0.0152498115500106])
# accel_init = np.array([9.31001853942871,	0.113502115011215,	-2.58657431602478])
# gyro_test = np.array([-0.0567961037158966,	0.0142987715080380,	0.0285795908421278])
# dt = 0.1
# a.initialize(accel_init, mag_init)
# a.time_update(gyro_test, dt)
# a.measurement_update(accel_init, mag_init)

File: matlab/ahrs_ekf.m
clear all
close all
clc

% columns 1-3 accel [m/s^2]
% 4-6 gyro [rad/s]
% 7-9 mag [uT]
% [data, dt] = loadSoftsysData("hw3_data.txt");

[data, dt, gps_course, gps_course_time] = loadNavData("data/normal_vecNav.mat");

% dt = 0.01;
% [data, true_orientation] = simImuData(dt);

accel = data(:, 1:3);
% gyro = deg2rad(data(:, 4:6));
gyro = data(:, 4:6);
mag = data(:, 7:9);

% mag calibration pre-processing step
[cal_data, ~, ~, ~] = loadNavData("data/6_28 Data/6-28_Mag_Cal_vecNav.mat");
mag_cal_data = data(:, 7:9);
[A, b, expmfs] = magcal(mag_cal_data);

% produce corrected mag data
mag = (mag - b)*A;

% initial rotation matrix, normalize columns
C_init = [cross(cross(accel(1,:)', mag(1,:)'), accel(1,:)') cross(accel(1,:)',mag(1,:)') accel(1,:)'];
C_init(:, 1) = C_init(:,1)./norm(C_init(:,1));
C_init(:, 2) = C_init(:,2)./norm(C_init(:,2));
C_init(:, 3) = C_init(:,3)./norm(C_init(:,3));

gyro_int(:,1) = [deg2rad(-90); deg2rad(50); deg2rad(-50)];

q_hat(:, 1) = [(1/2)*sqrt(C_init(1,1)+C_init(2,2)+C_init(3,3)+1);
               (1/2)*sign(C_init(3,2)-C_init(2,3))*sqrt(C_init(1,1)-C_init(2,2)-C_init(3,3)+1);
               (1/2)*sign(C_init(1,3)-C_init(3,1))*sqrt(C_init(2,2)-C_init(3,3)-C_init(1,1)+1);
               (1/2)*sign(C_init(2,1)-C_init(1,2))*sqrt(C_init(3,3)-C_init(1,1)-C_init(2,2)+1)];
q_hat_gyro_only(:, 1) = q_hat(:,1);
eul_gyro_only(:, 1) = [0; 0; 0];
eul(:, 1) = [0; 0; 0];
t(1) = 0;

P = 10*eye(4);
sigma_gyro = 0.075; % assuming equal noise on each axis
Sigma_gyro = (sigma_gyro^2).*eye(3);

% Q1 = (sigma_gyro^2).*eye(4);
% Q = Q1;

% assuming equal noise on each axis
sigma_acc = 0.5; 
sigma_mag = 1.1;
R = [(sigma_acc^2).*eye(3) zeros(3);
     zeros(3) (sigma_mag^2).*eye(3)];

magnetic_dip_angle = deg2rad(-14.78); % deg, for auburn, al, from magnetic-declination.com
magnetic_field_vec = [cos(magnetic_dip_angle) 0 sin(magnetic_dip_angle)]'./sqrt(cos(magnetic_dip_angle)^2 + sin(magnetic_dip_angle)^2); % ned
grav_vec = [0 0 -1]'; % ned

% magnetic_field_vec = [0 cos(magnetic_dip_angle) -sin(magnetic_dip_angle)]'./sqrt(cos(magnetic_dip_angle)^2 + sin(magnetic_dip_angle)^2); % enu
% grav_vec = [0 0 1]'; % enu

for i = 1:length(accel(:, 1))-1
    t(i+1) = dt*i;

    % --- time update----
    Omega = [0  -gyro(i+1,:);
             gyro(i+1,:)' -vec3ToSkewSym(gyro(i+1,:))];

    % from stack overflow - integrating quaternion using angular rates
    q_w_a = cos(norm(gyro(i+1, :))*dt/2);
    q_w_v = sin(norm(gyro(i+1, :))*dt/2).*gyro(i+1,:)'/(norm(gyro(i+1,:)));
    q_w = [q_w_a; q_w_v];

    % state_prop = cos(norm(gyro(i+1, :))*dt/2)*eye(4) + (2/(norm(gyro(i+1,:))))...
    %                 *sin(norm(gyro(i+1,:))*dt/2)*Omega;

    gyro_int(:,i+1) = gyro_int(:, i) + gyro(i+1, :)'*dt;

    % q_hat(:, i+1) = state_prop*q_hat(:,i);
    q_hat(:, i+1) = quatmult(q_hat(:,i), q_w);

    % normalize quaternion
    q_hat(:, i+1) = q_hat(:, i+1)./norm(q_hat(:, i+1));

    % q_hat_gyro_only(:, i+1) = state_prop*q_hat_gyro_only(:, i);
    q_hat_gyro_only(:, i+1) = quatmult(q_hat_gyro_only(:,i), q_w);


    % normalize quaternion
    q_hat_gyro_only(:, i+1) = q_hat_gyro_only(:, i+1)./norm(q_hat_gyro_only(:, i+1));    
    eul_gyro_only(:, i+1) = rad2deg(unwrap(quat2eul(q_hat_gyro_only(:, i+1)')));


    A_d = [1 (-dt/2)*gyro(i+1,1) (-dt/2)*gyro(i+1,2) (-dt/2)*gyro(i+1,3);
           (dt/2)*gyro(i+1,1) 1 (dt/2)*gyro(i+1,3) (-dt/2)*gyro(i+1,2);
           (dt/2)*gyro(i+1,2) (-dt/2)*gyro(i+1,3) 1 (dt/2)*gyro(i+1,1);
           (dt/2)*gyro(i+1,3) (dt/2)*gyro(i+1,2) (-dt/2)*gyro(i+1,1) 1;];

    % Jacobian of f w.r.t. angular rates (omega)
    W = (dt/2).*[-q_hat(2, i+1)  -q_hat(3, i+1)  -q_hat(4, i+1);
                  q_hat(1, i+1)  -q_hat(4, i+1)   q_hat(3, i+1);
                  q_hat(4, i+1)   q_hat(1, i+1)  -q_hat(2, i+1);
                 -q_hat(3, i+1)   q_hat(2, i+1)  -q_hat(1, i+1)];

    Q = W*Sigma_gyro*W'; % Q changes over time depending on angular vel
    
    % covariance propagation
    P = A_d*P*A_d' + Q;

    % --- measurement correction ---

    % normalize the accel and mag so we can compare them with our gravity
    % and magnetic field unit vectors
    accel_normalized = accel(i+1,:)./(norm(accel(i+1,:)));
    mag_normalized = mag(i+1,:)./(norm(mag(i+1,:)));

    % rotate gravity and mag vectors into sensor's frame (this is the
    % nonlinear measurement model, h(x))
    % rot_mat = quatToRotMat(q_hat(:, i+1));
    rot_mat = quat2rotm(q_hat(:,i+1)');
    accel_hat = rot_mat'*grav_vec;
    mag_hat = rot_mat'*magnetic_field_vec;
    Y_hat = [accel_hat; mag_hat];

    % form jacobian of measurement model (H)
    u_g = cross(grav_vec, q_hat(2:4, i+1));
    u_r = cross(magnetic_field_vec, q_hat(2:4, i+1));
    skew_sym_first_row = vec3ToSkewSym(u_g+q_hat(1,i+1)*grav_vec);
    skew_sym_second_row = vec3ToSkewSym(u_r+q_hat(1,i+1)*magnetic_field_vec);
    H = 2.*[u_g skew_sym_first_row+dot(q_hat(2:4,i+1),grav_vec)*eye(3)-grav_vec*q_hat(2:4,i+1)';
            u_r skew_sym_second_row+dot(q_hat(2:4,i+1),magnetic_field_vec)*eye(3)-magnetic_field_vec*q_hat(2:4,i+1)';
            ];

    Y = [accel_normalized'; mag_normalized'];

    % compute Kalman gain
    v = Y - Y_hat;
    S = H*P*H' + R;
    K = P*H'*inv(S);

    % apply measurement correction to state and covariance
    corrected_state = q_hat(:, i+1) + K*v;
    q_hat(:, i+1) = corrected_state(1:4);
    P = (eye(4) - K*H)*P';

    % normalize quaternion
    q_hat(:, i+1) = q_hat(:, i+1)./norm(q_hat(:, i+1));

    eul(:, i+1) = rad2deg(unwrap(quat2eul(q_hat(:, i+1)')));
end


% figure
% plot(t, eul(1, :));
% hold on
% plot(t, eul_gyro_only(1,:));
% % plot(t(1:end-1), true_orientation(:,1));
% plot(t, rad2deg(gyro_int(3,:)));
% legend("EKF", "Gyro Only", "Gyro Only (Euler)");
% title("Rotation about North Axis");
% 
% figure
% plot(t, eul(2, :));
% hold on
% plot(t, eul_gyro_only(2,:));
% % plot(t(1:end-1), true_orientation(:,2));
% plot(t, rad2deg(gyro_int(2,:)));
% legend("EKF", "Gyro Only", "Gyro Only (Euler)");
% title("Rotation about East Axis");

figure
plot(t, eul(3, :));
hold on
% plot(t, eul_gyro_only(1,:));
% plot(t(1:end-1), true_orientation(:,3));
plot(gps_course_time, rad2deg(gps_course));
% plot(t, rad2deg(gyro_int(1,:)));
legend("EKF", "GPS Course");
title("Heading");
